File: zenoh/examples/zenoh-net/zn_pub.rs
//
// Copyright (c) 2017, 2020 ADLINK Technology Inc.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
// which is available at https://www.apache.org/licenses/LICENSE-2.0.
//
// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
//
// Contributors:
//   ADLINK zenoh team, <zenoh@adlink-labs.tech>
//
use clap::{App, Arg};
use zenoh::net::*;

#[async_std::main]
async fn main() {
    // initiate logging
    env_logger::init();

    let args = App::new("zenoh-net pub example")
        .arg(Arg::from_usage("-m, --mode=[MODE] 'The zenoh session mode.")
            .possible_values(&["peer", "client"]).default_value("peer"))
        .arg(Arg::from_usage("-e, --peer=[LOCATOR]...  'Peer locators used to initiate the zenoh session.'"))
        .arg(Arg::from_usage("-p, --path=[PATH]        'The name of the resource to publish.'")
            .default_value("/demo/example/zenoh-rs-eval"))
        .arg(Arg::from_usage("-v, --value=[VALUE]      'The value of the resource to publish.'")
            .default_value("Pub from Rust!"))
        .get_matches();

    let config = Config::new(args.value_of("mode").unwrap()).unwrap()
        .add_peers(args.values_of("peer").map(|p| p.collect()).or_else(|| Some(vec![])).unwrap());
    let path    = args.value_of("path").unwrap();
    let value   = args.value_of("value").unwrap();

    println!("Openning session...");
    let session = open(config, None).await.unwrap();
    
    print!("Declaring Resource {}", path);
    let rid = session.declare_resource(&path.into()).await.unwrap();
    println!(" => RId {}", rid);
    
    println!("Declaring Publisher on {}", rid);
    let publ = session.declare_publisher(&rid.into()).await.unwrap();
    
    println!("Writing Data ('{}': '{}')...\n", rid, value);
    session.write(&rid.into(), value.as_bytes().into()).await.unwrap();

    session.undeclare_publisher(publ).await.unwrap();
    session.close().await.unwrap();
}

File: zenoh-ffi/examples/zn_pub.c
/*
 * Copyright (c) 2017, 2020 ADLINK Technology Inc.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
 * which is available at https://www.apache.org/licenses/LICENSE-2.0.
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
 *
 * Contributors:
 *   ADLINK zenoh team, <zenoh@adlink-labs.tech>
 */
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include "zenoh-ffi.h"

int main(int argc, char **argv) {
    char *uri = "/demo/example/zenoh-c-pub";
    if (argc > 1) {
        uri = argv[1];
    }
    char *value = "Pub from C!";
    if (argc > 2) {
        value = argv[2];
    }
    char *locator = 0;
    if (argc > 3) {
        locator = argv[3];
    }

    printf("Openning session...\n");
    ZNSession *s = zn_open(PEER, locator, 0);
    if (s == 0) {
        printf("Unable to open session!\n");
        exit(-1);
    }

    // long rid = zn_declare_resource(s, uri);

    // printf("Declaring Publisher on '%s'...\n", uri);
    // ZNPublisher *pub = zn_declare_publisher(s, uri);
    // if (sub == 0) {
    //     printf("Unable to declare publisher.\n");
    //     exit(-1);
    // }

    char buf[256];
    for(int idx = 0; 1; ++idx) {
        sleep(1);
        sprintf(buf, "[%4d] %s", idx, value);
        printf("Writing Data ('%s': '%s')...\n", uri, buf);
        zn_write(s, uri, buf, strlen(buf));
    }

    // zn_undeclare_publisher(pub);
    zn_close(s);
    return 0;
}